### ISSUE TITLE
Sending message to specific client by id and http request

### ISSUE THREAD

**Issue by mnami** [SocketCluster/socketcluster#57]
hi
my pushnotification server and application server are seprated, i have a web method like http://server/sendnotification/?clientid=hdfhjshkweelkja-564dff&message="test message"
my issue is how i can send this message to client is this id is exist?

**Comment by jondubois** [comment]
You will need to get your SC instance integrated with an external message queue like Redis or RabbitMQ. See http://github.com/topcloud/sc-redis

**Comment by mnami** [comment]
hi, thanks for your response, According to sc-redis  I cant access to socket object with socket id, and I can to publish message to channels, I need to access socket with socket id

**Comment by jondubois** [comment]
@mnami I'm not sure what your specific use case is but you generally shouldn't try to reference individual sockets - It's not a scalable approach - What if you have multiple servers and the socket you want is hosted on a different machine? It would be physically impossible to get a reference to it from the current machine.

^ These kinds of issues happen when you try to 'tightly couple' two specific client sockets together.
Sockets should always stay loosely coupled from each other and communicate through channels instead.

As a rule of thumb - Any kind of user-to-user communication should be done through pub/sub methods like publish() and subscribe() and watch() (you can call all of these from the client).

emit() and on() should only be used when you want user-to-server or server-to-user communication. For example, if the user wants to request a particular piece of data from the server, it could emit an event and the server would emit back the response containing the data.

**Comment by mnami** [comment]
i need to implement a http method to send a message to connected client, i know socket and its id saves in stores, your mean is i cant access to socket object with socket id?

**Comment by delaneyj** [comment]
@mnami if you know the socketId why not subscribe them to a 'user:{socketId}:{guid}' channel?  That way you can send individual things to them and still have pretty good guarantees for not spamming a clients channel.

**Comment by mnami** [comment]
hi, me again.
i want to add proxy like load balancer in previous version of sc, but initiate of workers is diffrent in new sc, can you give me a guidance about add proxy to route http request to specific process?
thanks

**Comment by mnami** [comment]
i must to say, old sc load balancer has this run time error:
Origin: Balancer
[Error] Error: bind EADDRINUSE
at exports._errnoException (util.js:746:11)
at cb (net.js:1178:33)
at shared (cluster.js:586:5)
at Worker.<anonymous> (cluster.js:561:9)
at process.<anonymous> (cluster.js:692:8)
at process.emit (events.js:129:20)
at handleMessage (child_process.js:324:10)
at child_process.js:408:7
at process.handleConversion.net.Native.got (child
at process.<anonymous> (child_process.js:407:13)

**Comment by jondubois** [comment]
@mnami Why do you want to add a proxy? You no longer have to add a local proxy to do load balancing between worker processes in SC2 thanks to its pure WebSocket implementation.

WebSockets never close the connection at any time so a client will stay bound to the same process on the same server (so you don't need session stickiness anymore - That was the whole point of having load balancers in SC1).

I don't know why you're getting EADDRINUSE, I can't reproduce this. It means that another process is using this port. Are you using --debug, --debug-workers or --debug-stores on CLI?

**Comment by mnami** [comment]
i know that websocket is connection oriented and i want to route some http request to known process.

**Comment by jondubois** [comment]
@mnami It was a convenient side effect of SC1 that HTTP requests from one user always ended up in the same process (HTTP session stickiness).

HTTP session stickiness is something SC2 tries to move away from because of these reasons:

- It's very expensive to do HTTP session stickiness - Based on some tests I did, SC2 is more than 10 times faster than SC1 when it comes to handling new connections (this is in part because the extra layer of proxies/loadbalancers required in SC1 is very expensive).
- HTTP is designed to be stateless. Trying to force it to be stateful (using session stickiness) introduces scalability/security challenges. For example, it's better if you don't allow clients to choose which host/process will handle their request (which is a requirement for stickiness) - Otherwise a malicious user might be able to target a specific process and launch a Denial of Service attack against that process. As soon as you remove HTTP session affinity/stickiness, this problem goes away.
- WebSocket is naturally stateful and is not prone to denial of service attacks like with sticky-session HTTP. When you need to handle stateful, sequential interactions between a client and a server, you should be using WebSockets (and not HTTP).

So in summary, you shouldn't rely on stickiness for HTTP because it's not the most secure approach at scale. I'm not sure what your use case is, but can you find a way to not rely on stickiness for HTTP - Maybe by storing data in a database (or using the scServer.global object)?

**Comment by MichaelJCole** [comment]
@jondubois hey, your first comment should really be in the README.  The designed uses aren't clear from the documentation.

**Comment by happilymarrieddad** [comment]
@damirci Hey man, our clients all subscribe to their own channel based on their id. You can then query data or whatever on the server and send the data back to them I guess. That being said, you can pass in a callback from any request client-side which is what will be called as the second argument on the server channel being listened to... if that makes sense...

**Comment by jondubois** [comment]
@MichaelJCole Yes sure, I'll try to add something along those lines in the main README. It's not a hard rule, but it's a good pattern to follow if you want your system to scale easily if/when the time comes.
